#441 WIP Source mongodb implementation

FUT opened the pull request:
## What
I've implemented a basic Mongodb implementation. The work is still in progress, but I decided to create a PR to ask several questions.

Closes #441 

## Pre-merge Checklist
- [x] *Run integration tests*
- [x] *Publish Docker images* `docker pull narrativebi/airbyte-source-mongodb:0.1.3`
- [x] Update SUMMARY.md to include the docs

## Recommended reading order
1. `airbyte-integrations/connectors/source-mongodb/source.rb`
2. `airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb`
3. the rest

## Follow up items (after this PR) 
- [ ] Create issue for autogenerating airbyte protocol ruby types in gradle CI
- [x] Create issue for running integration tests against a docker container

michel-tricot commented:
Hey @FUT, thank you so much for your PR.

We will look at it on Monday and try to understand why the integration tests are not working for you.

FUT commented:
I finally figured out what was wrong. I've also added errors logging to `airbyte-json-validation/src/main/java/io/airbyte/validation/json/JsonSchemaValidator.java`. Java is not my main language, please let me know if I should rework logging a bit.

FUT commented:
I've updated the main message, and pushed the docker image

FUT commented:
What is the best way to put documentation to page like https://docs.airbyte.io/integrations/sources/mongodb ?

FUT commented:
I plan to implement incremental sync for this connector in the future. What approach is preferable:

michel-tricot commented:
> What is the best way to put documentation to page like https://docs.airbyte.io/integrations/sources/mongodb ?

You can actually put an `md` file in `docs/integrations/sources`. I suggest you inspire yourself from an existing one so you don't create one scratch.

> I plan to implement incremental sync for this connector in the future.

@sherifnada what do you think? I like the idea of starting with the full sync and doing incremental as a subsequent merge.

michel-tricot commented:
Do you know if mongo as a concept of `updated_at` for top level records?

sherifnada commented:
Agreed that merging full refresh into master then creating a separate PR for incremental is better. It provides incremental (ha-ha) value and gives you @FUT a better idea of iterating within the codebase which is helpful in case you want to make any future changes. It also makes it easier to review!

I can review the changes here later today.

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb:
```diff
@@ -0,0 +1,81 @@
+  private
+
+  def client
+    if @client
```
might be simpler to do `if not @client` and run the else block, then always return `@client`

FUT replied:
I've reworked that memoization pattern in a bit different way. Please let me know if that approach will work

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_stream.rb:
```diff
@@ -0,0 +1,85 @@
+        if type != NilClass
+          map_type(type)
+        end
+      end.compact
```
is there a chance of getting nil in this `map`? I'm a little scared seeing the `compact` because it tells me that some columns might have a nil type and we are silently sweeping it under the rug.

FUT replied:
Yes, there is. MongoDB does not have anything like table definition, thus we have to define column types from actual attribute values. When we discover schema, we use only the first 10k records in each collection. There is a chance that we encounter 10k `null` values and can not define column type properly. In that case we fallback to the `string` type and rely on custom data transformation of imported data.

FUT replied:
Finally, I decided to go a different way. I reworked the `discover` command to run the map-reduce command which should be sufficient even for large clusters and will return all unique document props in the collection. Then trying to determine prop type selecting 10k records from the collection where this prop is not empty.

I believe that approach will work for all cases. Nevertheless, I will update the readme

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_reader.rb:
```diff
@@ -0,0 +1,81 @@
+  end
+
+  def read
+    @catalog['streams'].each do |stream_object|
```
a better name would be `configured_stream` since that is the name used in the JSON schema

FUT replied:
In fact `ConfiguredAirbyteCatalog` has a single `streams` key which is an array of `ConfiguredAirbyteStream` objects.

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_reader.rb:
```diff
@@ -0,0 +1,81 @@
+  def read
+    @catalog['streams'].each do |stream_object|
+      stream = stream_object['stream']
+      AirbyteLogger.log("Reading stream #{stream['name']} in #{stream['sync_mode']} mode")
```
`sync_mode` should be a field on `stream_object`. See [`ConfiguredAirbyteStream`](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L158)

FUT replied:
I've changed the naming of local variables to make it super clear. You are right, `AirbyteStream` has `supported_sync_modes` and `ConfiguredAirbyteStream` has `sync_mode`. Here we are using a `ConfiguredAirbyteStream` object which has the `sync_mode` attribute.

FUT commented:
> Incremental: Do you know if mongo as a concept of `updated_at` for top level records?
I believe timestamps behavior is ORM-specific. I see that mongoose has [the option](https://mongoosejs.com/docs/guide.html#timestamps) to autogenerate nested parameter `{..., timestamps: {createdAt: XXX, updatedAt: XXX}}`. On the other hand, MongoID has [the option](https://docs.mongodb.com/mongoid/master/tutorials/mongoid-documents/#timestamping) to generate plan attributes `{..., createdAt: XXX, updatedAt: XXX}`.

sherifnada reviewed (approved):
Looks great! Are standard tests succeeding for you now? 

How are you setting up the custom tests? are you hitting a pre-configured mongodb instance somewhere online?

sherifnada commented on airbyte-integrations/connectors/source-mongodb/README.md:
```diff
@@ -0,0 +1,20 @@
+# Mongodb Source 
+
+This is the repository for the Mongodb source connector, written in Ruby. 
```
would you like to add yourself under an author header? ;)

FUT replied:
Why not :). But unfortunately, I did not find anything similar in other connectors. Could you please give me an example of what should I write there?

sherifnada replied:
added a suggestion below. Will review the rest of the changes later!

FUT commented on airbyte-json-validation/src/main/java/io/airbyte/validation/json/JsonSchemaValidator.java:
```diff
@@ -52,8 +52,17 @@ public JsonSchemaValidator() {
+      .validate(objectJson);
+
+    if (!validationMessages.isEmpty()) {
+      String str = String.format(
```
That's a good point. I decided to move that logging to `test` method since we are technically losing information about errors there.

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/airbyte_protocol.rb:
```diff
@@ -0,0 +1,461 @@
+# The file was generated in several steps.
```
I thought about that, and even tried, but I am new to `gradle` and it turned out to be a pretty complex task

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb:
```diff
@@ -0,0 +1,81 @@
+      result = {'status' => Status::Succeeded}
+    rescue Mongo::Auth::Unauthorized => e
+      AirbyteLogger.log(e.backtrace.join("\n"), Level::Fatal)
+      result = {'status' => Status::Failed, 'message' => 'Authentication failed.'}
```
Unfortunately, it does not. All we get is auth failed error. Nevertheless, I've applied formatter to Mongo driver logger. So all messages from Mongo driver will also be properly formatted and will get to worker logs. So, we have maximum transparency here

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb:
```diff
@@ -0,0 +1,81 @@
+    rescue Mongo::Auth::Unauthorized => e
+      AirbyteLogger.log(e.backtrace.join("\n"), Level::Fatal)
+      result = {'status' => Status::Failed, 'message' => 'Authentication failed.'}
+    end
```
Sounds reasonable. Fixed that.

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb:
```diff
@@ -0,0 +1,81 @@
+  def check(config:)
+    @config = JSON.parse(File.read(config))
+
+    result = {'status' => Status::Failed, 'message' => 'Something went wrong.'}
```
I removed this line at all since we catch all errors now

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb:
```diff
@@ -0,0 +1,81 @@
+    if @client
+      @client
+    else
+      uri = "mongodb://#{@config['user']}:#{@config['password']}@#{@config['host']}:#{@config['port']}/#{@config['database']}?authSource=admin"
```
I assume there might be a very rare situation when authentication is made against another database. I will add authSource to config with default value admin

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_stream.rb:
```diff
@@ -0,0 +1,85 @@
+    TrueClass => AIRBYTE_TYPES[:boolean],
+    FalseClass => AIRBYTE_TYPES[:boolean],
+    Array => AIRBYTE_TYPES[:array],
+    Hash => AIRBYTE_TYPES[:hash],
```
You are right, that's a typo :) Fixed that

FUT commented:
> Are standard tests succeeding for you now? 
> How are you setting up the custom tests? are you hitting a pre-configured mongodb instance somewhere online?
Yes, they are. I've added docs to README for test MongoDB configuration

I've addressed all the comments (thank you for the review!) and added documentation for that source to `docs` folder.

FUT commented:
By the way, what would you recommend to read (which connector or doc) about incremental sync? I understand the general idea, but a specific example of implementation will help a lot. Toy connector docs helped me a lot with full sync but they are missing the incremental sync topic.

sherifnada commented on airbyte-integrations/connectors/source-mongodb/README.md:
```diff
@@ -0,0 +1,49 @@
+# Mongodb Source 
+
+This is the repository for the Mongodb source connector, written in Ruby. 
+
```
```suggestion
## Author
This connector was authored by [Yury Kaliada](github.com/FUT).
```

FUT replied:
Thank you!

FUT commented:
I've also implemented incremental sync mode :)

FUT commented:
I am not sure why do I get such an invalid UI after the discovery phase:

![Airbyte](https://user-images.githubusercontent.com/965875/109004372-c0c38b00-76b9-11eb-91a8-bd5451328948.png)

That prevents me from selecting some specific column for incremental sync. Discover JSON looks like valid for me:

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_configured_stream/incremental.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def cursor_field
+    @cursor_field ||= configured_stream['cursor_field']&.first
```
the docs should indicate that the cursor field can't be nested (the Airbyte protocol allows nested cursor fields, that's why the field is an array)

FUT replied:
agree. added to the docs

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_state.rb:
```diff
@@ -0,0 +1,32 @@
+
+    AirbyteLogger.log("Saving state:\n#{JSON.pretty_generate(@state)}")
+
+    File.write(state_file, @state.to_json)
```
should this be logging state to stdout as a state message?

FUT replied:
I've missed that somehow. Fixed now

sherifnada commented on airbyte-integrations/connectors/source-mongodb/secrets.sample/incremental_configured_catalog.json:
```diff
@@ -0,0 +1,69 @@
+{
+  "streams":[
+    {
+      "sync_mode": "full_refresh",
```
for the tests to correctly verify incremental functionality this should be set to `incremental` with the correct cursor field

FUT replied:
Changed one stream to incremental

sherifnada commented on airbyte-json-validation/src/main/java/io/airbyte/validation/json/JsonSchemaValidator.java:
```diff
@@ -57,7 +57,16 @@ public JsonSchemaValidator() {
+      String str = String.format(
+          "JSON schema validation failed. \nerrors: %s",
+          Strings.join(validationMessages, ", "));
+      System.out.println(str);
```
For logging, it's best to use a `Logger` object. Typically we declare a `LOGGER` at the top of the file and use something like `LOGGER.info(..)`. Here are examples: 

* [Defining the logger](https://github.com/airbytehq/airbyte/commit/38e84f1746b4c35b92dd646514cf67921badb7e1#diff-b5220bb6ab91f624515384c9c4a153c5116176cc79eb7f0835029bd326736883R35)
* [Using the logger](https://github.com/airbytehq/airbyte/commit/38e84f1746b4c35b92dd646514cf67921badb7e1#diff-b5220bb6ab91f624515384c9c4a153c5116176cc79eb7f0835029bd326736883R55)

FUT replied:
Fixed that

sherifnada commented on docs/integrations/sources/mongodb.md:
```diff
@@ -0,0 +1,67 @@
+# Mongodb 
+
+The MongoDB source supports Full Refresh only.
```
we should update docs for incremental

FUT replied:
:+1:

sherifnada commented on docs/integrations/sources/mongodb.md:
```diff
@@ -0,0 +1,67 @@
+The MongoDB source supports Full Refresh only.
+
+## Resulting schema 
+MongoDB does not have anything like table definition, thus we have to define column types from actual attributes and their values. Discover phase have two steps:
```
```suggestion
MongoDB does not have anything like table definition, thus we have to define column types from actual attributes and their values. The discover phase has two steps:
```

sherifnada commented on docs/integrations/sources/mongodb.md:
```diff
@@ -0,0 +1,67 @@
+MongoDB does not have anything like table definition, thus we have to define column types from actual attributes and their values. Discover phase have two steps:
+
+### Step 1. Find all unique properties
+Connector runs the map-reduce command which returns all unique document props in the collection. Map-reduce approach should be sufficient even for large clusters.
```
```suggestion
Connector runs the [map-reduce command](https://docs.mongodb.com/manual/core/map-reduce/)  which returns all unique document props in the collection. Map-reduce approach should be sufficient even for large clusters.
```

sherifnada commented on docs/integrations/sources/mongodb.md:
```diff
@@ -0,0 +1,67 @@
+Connector runs the map-reduce command which returns all unique document props in the collection. Map-reduce approach should be sufficient even for large clusters.
+
+### Step 2. Determine property types
+For each property found, connector selects 10k documents from the collection where this property is not empty. If all the selected values have the same type - connector will set appropriate type to the property. In all other cases connector will fallback to `string` type.
```
```suggestion
For each property found, the connector selects 10k documents from the collection where this property is not empty. If all the selected values have the same type - connector will set appropriate type to the property. In all other cases connector will fallback to `string` type.
```

sherifnada commented on docs/integrations/sources/mongodb.md:
```diff
@@ -0,0 +1,67 @@
+
+Run `mongo` shell, switch to `admin` database and create three users: 
+
+1. ADMIN_USER will be used for database administration (you can skip this user in case you already use authentication for MongoDB access).
```
I think Airbyte's instructions should only describe the read only user case because presumably someone syncing data from Mongo already has a mongo instance setup, so we shouldn't give instructions about administrating the DB or interacting with it from their application. The scope of this guide is larger/requires more maintenance if we give instructions for maintaining the Mongo database instance. WDYT?

FUT replied:
That's reasonable. I've narrowed the scope of instructions. Please let me know if you think I should drop even more information.

sherifnada commented:
@FUT probably the easiest implementation to reference for incremental is the [mailchimp source](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-mailchimp/source_mailchimp/client.py#L61) 

You have most of the implementation correct as far as I can tell but just need to emit state messages

sherifnada commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_stream.rb:
```diff
@@ -0,0 +1,88 @@
+
+    AirbyteStream.from_dynamic!({
+      "name" => @collection.name,
+      "supported_sync_modes" => [SyncMode::FullRefresh, SyncMode::Incremental],
```
```suggestion
      "supported_sync_modes" => [SyncMode::FullRefresh, SyncMode::Incremental],
      "source_declared_cursor" => false,
```

sherifnada commented on airbyte-integrations/connectors/source-mongodb/README.md:
````diff
@@ -0,0 +1,52 @@
+docker build . -t airbyte/source-mongodb:dev
+```
+
+### Integration Tests 
````
PTAL: https://github.com/narrative-bi/airbyte/pull/1

FUT replied:
I've fixed the `Dockerfile.test` and added small dumps here

FUT commented on airbyte-integrations/connectors/source-mongodb/lib/mongodb_configured_stream/incremental.rb:
```diff
@@ -0,0 +1,82 @@
+
+  private
+
+  # Rely on a descriptive naming for type definition. It's too expensive to check the cursor type of every document in advance.
```
What do you think about that solution? I do not see any other inexpensive way to determine the cursor type.

sherifnada replied:
what if we do something like what we do in [`discover_property_type`](https://github.com/airbytehq/airbyte/pull/2125/files#diff-f28873de3c945904b586e78fa4a1bb1214aa57425680b79206f359984d49734cR67) when the stream is initialized instead? This way we only make a query for say, 1k documents, and determine the type based on that.

sherifnada reviewed (commented):
Almost there! 🎉

FUT commented on airbyte-integrations/connectors/source-mongodb/README.md:
````diff
@@ -0,0 +1,51 @@
+```
+
+### Integration Tests 
+1. Setup MongoDB on your local machine
````
That's right. I've removed that

FUT commented:
I reworked type determination and moved it to `airbyte-integrations/connectors/source-mongodb/lib/mongodb_types_explorer.rb`. 

Tests are working properly :)

FUT commented:
I fixed merge conflict and rebased to master branch. Could you please check once again?

sherifnada reviewed (approved):
LGTM! Will merge this tomorrow. Thank you so much for contributing this source -- very excited to have it running in production!!

sherifnada commented:
@FUT one last thing is needed: can you update [`SUMMARY.md`](https://github.com/airbytehq/airbyte/blob/master/docs/SUMMARY.md) file to point to your docs under the `Integrations>Sources`  section?
